gtsam/hybrid/MixtureFactor.h

```diff
@@ -107,8 +107,12 @@ class MixtureFactor : public HybridFactor {
       std::copy(f->keys().begin(), f->keys().end(),
                 std::inserter(factor_keys_set, factor_keys_set.end()));
 
-      nonlinear_factors.push_back(
-          boost::dynamic_pointer_cast<NonlinearFactor>(f));
+      if (auto nf = boost::dynamic_pointer_cast<NonlinearFactor>(f)) {
```

Comment:
Why is this templated? Can't we just have boost::shared_ptr<NonlinearFactor> anyway? which, BTW, is known as a `NonlinearFactorGraph` ? Then the implementation can also be moved to cpp.

PS, hoping the above is a rhetorical question but maybe there is a reason I don't understand.

Reply:
The reason is that boost shared_pointers don't seem to allow type hierarchy. For example, if I remove the template, this works

```cpp
using MotionModel = BetweenFactor<double>;

auto noise_model = noiseModel::Isotropic::Sigma(1, 1.0);
auto still = boost::make_shared<MotionModel>(X(0), X(1), 0.0, noise_model),
     moving = boost::make_shared<MotionModel>(X(0), X(1), 1.0, noise_model);

std::vector<NonlinearFactor::shared_ptr> components = {still, moving};
auto dcFactor = boost::make_shared<MixtureFactor>(
    contKeys, DiscreteKeys{gtsam::DiscreteKey(M(1), 2)}, components);
```

but this doesn't
```cpp
using MotionModel = BetweenFactor<double>;

auto noise_model = noiseModel::Isotropic::Sigma(1, 1.0);
auto still = boost::make_shared<MotionModel>(X(0), X(1), 0.0, noise_model),
     moving = boost::make_shared<MotionModel>(X(0), X(1), 1.0, noise_model);

std::vector<MotionModel::shared_ptr> components = {still, moving};
auto dcFactor = boost::make_shared<MixtureFactor>(
    contKeys, DiscreteKeys{gtsam::DiscreteKey(M(1), 2)}, components);
```

The only difference is the type of `components`. 😕

Reply:
I guess this becomes a matter of API convenience. Being able to pass in `vector<SpecificType>` rather than `vector<NonlinearFactor>` seems cleaner to me. Same reason for not using `NonlinearFactorGraph` even though they are semantically the same. I can update this if you'd still like.

Reply:
Also, I can't do this little neat trick: `NonlinearFactorGraph components = {still, moving};`. Maybe this indicates something missing from the `FactorGraph` class.